### PR TITLE
[FLINK-20271][FLINK-20272][table] Fix ArrayIndexOutOfBounds and wrong result for TopN

### PR DESCRIPTION
**Pull request by wuchong**
## What is the purpose of the change

This pull request fixes 2 bugs in `RetractableTopNFunction`:
- FLINK-20271: ArrayIndexOutOfBoundsException when retracting a record in the TopN
- FLINK-20272: Wrong result when removing records out of TopN

## Brief change log

- FLINK-20271: Fix the `curRank` cursor in `retractRecordWithoutRowNumber`.
- FLINK-20272: Remove record from state if the record is out of TopN range.

## Verifying this change

- Added unit tests which can reproduce the reported exceptions.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by wuchong** [comment]
@fsk119 , could you help to review this?

**Review by fsk119 — COMMENTED** [review]
Thanks for your contribution. I left one comments.

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/rank/RetractableTopNFunction.java [hunk 1]
```diff
@@ -190,6 +191,19 @@ public void processElement(RowData input, Context ctx, Collector<RowData> out) t
 				}
 			}
 
+			if (!stateRemoved) {
+				// the input record has not been removed from state
+				// should update the data state
+				List<RowData> inputs = dataState.get(sortKey);
+				if (inputs != null) {
+					inputs.remove(input);
```
[the remark this answers is not in the document]
Comment (wuchong) [reply flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/rank/RetractableTopNFunction.java:199]: Good point.

**Review by fsk119 — APPROVED** [review]
LGTM

**Pull request closed without merging**